AMP ESP32 support fixes and improvements

marekmatej opened the pull request:
Supporting change to fix and update the IPM samples on ESP32.

marekmatej committed:
boards: esp32: update dts for APPCPU

Use ready made partition tables on ESP32 targets.
Add flash slot partition for APPCPU.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
samples: driver: ipm: esp32 sysbuild refactor

Make APP_CPU board name auto generated and remove the related config.
Remove the boards overlay and replace with a socs overlays.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32: fixes and updates for AMP

Multiple AMP related updates:

- use common AMP Kconfig
- rework the APPCPU linker script
- use MCUboot image format for APPCPU image
- fix multi-processing startup code

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32s3: update APPCPU code

Updates for AMP targets

- use common AMP Kconfig
- update APPCPU linker script
- place AMP common area in the reserved space

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
drivers: clock_control: limit APPCPU clock setup

Update init function so APPCPU could not altere the clock setup.
Fix build in case if no inter-cpu module is selected.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
manifest: AMP supporting changes

Fetch supporting changes.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

nordicjm commented on soc/espressif/common/Kconfig.amp:
```diff
@@ -0,0 +1,26 @@
+	hex "ESP32 APPCPU DRAM size"
+	default 0x10000
+	help
+	  Defines APPCPU DRAM area size in bytes.
```
these should be in dts

marekmatej replied:
That would be the next step I'm afraid, as it doesn't fit to the current model.

nordicjm commented on soc/espressif/common/Kconfig.amp:
```diff
@@ -0,0 +1,26 @@
+	hex "ESP32 APPCPU IRAM size"
+	default 0x10000
+	help
+	  Defines APPCPU IRAM area size in bytes.
```
also confused, you add this file in this comment then instantly delete it in the next commit

marekmatej replied:
missed that during rebase, thx for spotting this!

nordicjm commented on samples/drivers/ipm/ipm_esp32/Kconfig.sysbuild:
```diff
@@ -1,11 +1,10 @@
-	default "esp32_devkitc_wrover/esp32/appcpu" if $(BOARD) = "esp32_devkitc_wroom"
-	default "esp32_devkitc_wroom/esp32/appcpu" if $(BOARD) = "esp32_devkitc_wroom"
-	default "esp32s3_devkitm/esp32s3/appcpu" if $(BOARD) = "esp32s3_devkitm"
+#
+#source "share/sysbuild/Kconfig"
+#
+#config IPM_REMOTE_BOARD
+#	string
+#	default "esp32_devkitc_wroom/esp32/appcpu" if $(BOARD) = "esp32_devkitc_wroom"
+#	default "esp32s3_devkitm/esp32s3/appcpu" if $(BOARD) = "esp32s3_devkitm"
```
whole file can go?

marekmatej replied:
Actually, commenting this was a mistake as I usually do manual builds instead of sysbuild.

raffarost commented on soc/espressif/common/loader.c:
```diff
@@ -264,7 +264,8 @@ void __start(void)
 #endif
 
-#if !defined(CONFIG_MCUBOOT) && !defined(CONFIG_SOC_ESP32S3_APPCPU)
+#if !defined(CONFIG_SOC_ESP32S3_APPCPU) && !defined(CONFIG_SOC_ESP32S3_APPCPU) && \
```
does this condition repeat CONFIG_SOC_ESP32S3_APPCPU twice?

marekmatej replied:
thx, should be esp32 && esp32s3. Fixed

raffarost commented on soc/espressif/esp32/memory.h:
```diff
@@ -5,18 +5,19 @@
-#define SRAM0_IRAM_START    0x40070000
-#define SRAM0_CACHE_SIZE    0x10000
-#define SRAM0_SIZE          0x30000
+#define SRAM0_START           0x40070000
```
just wondered if these settings are following the concept of PR #80340

marekmatej replied:
No, those need to be reworked if/when #80340 will be merged.

raffarost commented on soc/espressif/esp32/soc.c:
```diff
@@ -185,3 +132,8 @@ void sys_arch_reboot(int type)
+
+#if defined(CONFIG_SOC_ENABLE_APPCPU) && !defined(CONFIG_MCUBOOT)
+extern int esp_appcpu_init(void);
+SYS_INIT(esp_appcpu_init, APPLICATION, 50);
```
suggestion: use defines instead of hardcoded values

sylvioalves replied:
You could use some of the available priorities such as `KERNEL_INIT_PRIORITY_DEVICE` or `APPLICATION_INIT_PRIORITY` instead of the hard-coded one.

raffarost commented on soc/espressif/esp32s3/memory.h:
```diff
@@ -78,8 +78,8 @@
 #if defined(CONFIG_SOC_ENABLE_APPCPU) || defined(CONFIG_SOC_ESP32S3_APPCPU)
-#define APPCPU_IRAM_SIZE CONFIG_ESP32S3_APPCPU_IRAM_SIZE
-#define APPCPU_DRAM_SIZE CONFIG_ESP32S3_APPCPU_DRAM_SIZE
+#define APPCPU_IRAM_SIZE CONFIG_ESP32_APPCPU_IRAM_SIZE
```
is this really right? (changed from ESP32S3 to ESP32)

marekmatej replied:
yes, since the AMP kconfig is being moved to the `common` folder it is no longer SoC specific, therefore rename.

raffarost replied:
right. Naming is sometimes tricky, but I'd think about one that doesn't seem like a direct mention to a device.
Maybe ESP instead of ESP32

marekmatej replied:
Yes, that is moved to common folder which applies for all AMP capable targets. I'll take a look if we can use `ESP` instead of `ESP32`.

sylvioalves commented on soc/espressif/esp32/default_appcpu.ld:
```diff
@@ -104,10 +160,10 @@ SECTIONS
     /* This goes here, not at top of linker script, so addr2line finds it last,
        and uses it in preference to the first symbol in IRAM */
-    _iram_start = ABSOLUTE(0);
+  /*  _iram_start = ABSOLUTE(0); */
```
Will you remove this?

sylvioalves commented on soc/espressif/esp32/default_appcpu.ld:
```diff
@@ -332,38 +446,20 @@ SECTIONS
-  {
-    _stext = .;
-    _text_start = ABSOLUTE(.);
+  /* --- START OF IROM --- */
```
I think this comments with empty content is only to tell there isn't DROM neither IROM stuff? Perhaps this could also go.

LucasTambor commented on soc/espressif/esp32/CMakeLists.txt:
```diff
@@ -24,13 +26,24 @@ zephyr_library_sources_ifdef(CONFIG_POWEROFF poweroff.c)
+
+board_runner_args(esp32 "--esp-monitor-baud=${monitor_baud}")
+
+message("-- Espressif HAL path: ${ESP_IDF_PATH}")
```
seems like a debug leftover

marekmatej replied:
In fact, I put it there for a reason. Same line is in the S3 port.

marekmatej committed:
dts: espressif: Update AMP sram nodes for ESP32 and ESP32-S3

Set AMP dts nodes (ipm, mbox, ...) to use fixed locations in reserved
memory areas. Those areas memories are delimited in the `memory.h`.
Size of the occupied areas can be calculated but the dts nodes addresses
needs to be set manually in every case.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej commented:
Folks, I found some AMP memories were overlapping with ordinary code on ESP32. Similar in ESP32-S3 I found a better location for the AMP common memory so I fixed both issues. Please revisit and re-approve the PR.

raffarost commented on samples/drivers/ipm/ipm_esp32/src/procpu_shell.c:
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#if defined(CONFIG_IPM) || defined(CONIFG_MBOX)
```
CONIFG_MBOX -> CONFIG_MBOX
